File: C/Curso 2/forca.c
#include <stdio.h>
#include <string.h>
#include "forca.h" // Include com aspas servem para importarmos header files que estão presentes no computador e não no compilador
#include <time.h>
#include <stdlib.h>

void abertura() {
    printf("\n-----//-----\n");
    printf("Jogo de Forca\n");
    printf("-----//-----\n\n");
}

void fazer_chute(char chutes[26], int* chutes_dados){ // Para fazer com que a função receba um endereço de memória ao invés de uma variável, deve-se colocar o * depois do tipo da variável
    char chute;
    printf("Faça seu chute: ");
    scanf(" %c", &chute); // Por conta do ENTER poder ser considerado um CHAR, para ignorá-lo o  certo é colocar espaço %c
    printf("\n");

    chutes[*chutes_dados] = chute; // O * antes da variável faz com que acesse o conteúdo do endereço de memória
    (*chutes_dados)++;
}

int ja_chutou(char chutes[26], int chutes_dados, char letra){
    int achou = 0;

    for (int j=0; j< chutes_dados; j++){
        if (chutes[j] == letra){
            achou = 1;
            break;
        }
    }

    return achou;
}

void desenha_forca(char palavrasecreta[20], char chutes[26], int chutes_dados){
    int erros = chutes_errados(chutes_dados, palavrasecreta, chutes);

    printf("  _______       \n");
    printf(" |/      |      \n");
    printf(" |      %c%c%c  \n", (erros >=1 ? '(': ' '),(erros >=1 ? '_' : ' '),(erros >=1 ? ')' : ' ')); // Esse jeito de escrever o if é chamado de if ternário (True : False)
    printf(" |      %c%c%c  \n", (erros >=3 ? '/' : ' '),(erros >=2 ? '|' : ' '),(erros >=4 ? '\\' : ' '));
    printf(" |       %c     \n", (erros >=2 ? '|' : ' '));
    printf(" |      %c %c   \n", (erros >=5 ? '/' : ' '), (erros >=5 ? '\\' : ' '));
    printf(" |              \n");
    printf("_|___           \n");

    printf("\n\n");

    for (int i=0; i < strlen(palavrasecreta); i++){
        int achou = ja_chutou(chutes, chutes_dados, palavrasecreta[i]);

        if (!achou){
            printf("_ ");
        } else {
            printf("%c ", palavrasecreta[i]);
        }
    }
    printf("\n");
}

int chutes_errados(int chutes_dados, char palavrasecreta[20], char chutes[26]){
    int erros = 0;
    for (int i=0; i< chutes_dados; i++){
        int existe = 0;
        for (int j=0;j<strlen(palavrasecreta);j++) {
            if (palavrasecreta[j] == chutes[i]){
                existe = 1;
                break;
            }
        }
        if (!existe){
            erros++;
        }
    }
    return erros;
}

int ja_enforcou(int chutes_dados, char palavrasecreta[20], char chutes[26]) {
    int erros;
    erros = chutes_errados(chutes_dados, palavrasecreta, chutes);
    return erros >= 5;
}

int ja_ganhou(char palavrasecreta[20], int chutes_dados, char chutes[26]) {
    for (int i=0; i<strlen(palavrasecreta); i++){
        if (!ja_chutou(chutes, chutes_dados, palavrasecreta[i])) {
            return 0;
        }
    }
    return 1;
}

void escolhe_palavra(char palavrasecreta[20]) {
    FILE* file = fopen("palavras.txt", "r");
    if (!file) {
        printf("Arquivo não disponível\n\n");
        exit(1);
    }

    int quantidade_palavras;
    fscanf(file, "%d", &quantidade_palavras);

    srand(time(0));
    int numero_random = rand() % quantidade_palavras;

    for (int i = 0; i <= numero_random; i++) {
        fscanf(file, "%s", palavrasecreta);
    }

    fclose(file);
}


int main() {
    char palavrasecreta[20]; // Como não existe o tipo string em C, utilizamos um array de char para isso

    escolhe_palavra(palavrasecreta);

    int acertou = 0; 
    int enforcou = 0;
    char chutes[26]; // O array já é um ponteiro
    int chutes_dados = 0;

    abertura();
    
    do{
        desenha_forca(palavrasecreta, chutes, chutes_dados);

        fazer_chute(chutes, &chutes_dados); // Para utilizar o endereço de memória da variável, coloca-se o & antes dela

        enforcou = ja_enforcou(chutes_dados, palavrasecreta, chutes);

        acertou = ja_ganhou(palavrasecreta, chutes_dados, chutes);

    } while (!acertou && !enforcou);

    desenha_forca(palavrasecreta, chutes, chutes_dados);

    if (acertou) {
        printf("Parabéns, você ganhou!\n");
    } else {
        printf("Que pena, a palvra secreta era %s.\nTente novamente!\n", palavrasecreta);
    }

}

// Esse código está com os locais das variáveis dentro da função main() para o aprendizado sobre ponteiros poder ficar demonstrado, ao invés de apenas as utilizarmos como variáveis globais
// Para escrever em arquivos, utilizar a função fprintf(), e para posicionar o ponteiro dentro do arquivo utilizar fseek()
// outra função útil para a leitura de arquivos é a feof(), ela retorna se o ponteiro está no final do arquivo

File: PYTHON/Aula  Python 1/Forca.py
import random

def jogar() -> None:

    print('***************************')
    print('Bem-vindo ao jogo da forca!')
    print('***************************')

    arquivo = open("Palavras.txt", "r")
    palavras = list()
    for linha in arquivo:
        linha = linha.strip()
        linha = linha.upper()
        palavras.append(linha)

    arquivo.close()

    numero = random.randrange(0,len(palavras))
    palavra_secreta = palavras[numero]

    enforcou = False
    acertou = False
    palavra_jogador = list('_' for letra_secreta in palavra_secreta)
    erros = 0

    while not enforcou and not acertou:
        print(f'Erros:{erros}/7')
        posicao = 0

        print(''.join(palavra_jogador))

        chute = input('Letra:')
        chute = chute.upper()
        
        if chute in palavra_secreta:
            posicao = 0
            for letra in palavra_secreta:
                if chute == letra:
                    palavra_jogador[posicao] = letra
                posicao += 1
        else:
            erros += 1
            desenha_forca(erros)
        
        if erros == 7:
            print('Você perdeu')
            enforcou = True

        if not '_' in palavra_jogador:
            print('Você ganhou')
            acertou = True

    print('Fim de Jogo!')

if __name__ == "__main__":
    jogar()

def desenha_forca(erros):
    print("  _______     ")
    print(" |/      |    ")

    if(erros == 1):
        print(" |      (_)   ")
        print(" |            ")
        print(" |            ")
        print(" |            ")

    if(erros == 2):
        print(" |      (_)   ")
        print(" |      \     ")
        print(" |            ")
        print(" |            ")

    if(erros == 3):
        print(" |      (_)   ")
        print(" |      \|    ")
        print(" |            ")
        print(" |            ")

    if(erros == 4):
        print(" |      (_)   ")
        print(" |      \|/   ")
        print(" |            ")
        print(" |            ")

    if(erros == 5):
        print(" |      (_)   ")
        print(" |      \|/   ")
        print(" |       |    ")
        print(" |            ")

    if(erros == 6):
        print(" |      (_)   ")
        print(" |      \|/   ")
        print(" |       |    ")
        print(" |      /     ")

    if (erros == 7):
        print(" |      (_)   ")
        print(" |      \|/   ")
        print(" |       |    ")
        print(" |      / \   ")

    print(" |            ")
    print("_|___         ")
    print()